velox/docs/develop/aggregate-functions.rst

```diff
@@ -152,13 +152,33 @@ A simple aggregation function is implemented as a class as the following.
     using IntermediateType = Array<Generic<T1>>;
     using OutputType = Array<Generic<T1>>;
 
+    // Define a struct for function-level states. Even if the aggregation function
+    // doesn't use function-level states, it is still necessary to define an empty
+    // FunctionState struct.
+    struct FunctionState {
+      // Optional.
+      TypePtr resultType;
+    };
+
+    // Optional. Defined only when the aggregation function needs to use function-level states.
+    // This method is called once when the aggregation function is created.
+    static void initialize(
+        core::AggregationNode::Step step,
+        FunctionState& state,
+        const std::vector<TypePtr>& rawInputTypes,
+        const TypePtr& resultType,
+        const std::vector<VectorPtr>& constantInputs,
+        std::optional<core::AggregationNode::Step> companionStep) {
```

Comment:
Can you document all the parameters? What is 'companionStep'? It seems strange that we have such a parameter as function implementations should be agnostic to whether they are used as "regular" or a "companion" function. 

CC: @kagamiori

Reply:
For the partial companion function, we need to know that its companion step is `kPartial`, while the agg function itself includes `kPartial`, `kFinal`, or `kSingle` steps.

Reply:
@liujiayi771 Thank you for updating PR description. It sounds like we want to allow stateful simple aggregate functions. This makes sense, but I wonder if we can make it work similar to Simple Function API for scalar functions.  There, function author defines a struct with call-once initialize and call-per-row call methods. The author is then free to add member variables to hold state and initialize it however they want from 'initialize'. Would it make sense to follow this pattern for aggregate functions as well?


> The result type is currently not exposed to author-defined accumulator type in the simple function interface.

Can we expose this?

> Some functions perform heavy computation on a constant argument before
> processing all input rows, such as approx_most_frequent, and store the
> computation result in a function-level state.

Would you clarify what is "heavy computation" done by approx_most_frequent to help readers understand a bit better?

> function-level state.

I wonder if a more accurate term would be "per-instance state". There is only one function, foo, but there are many instances of 'foo' and each has its own state, no?

Reply:
Hi @mbasmanova,
> but I wonder if we can make it work similar to Simple Function API for scalar functions. There, function author defines a struct with call-once initialize and call-per-row call methods. The author is then free to add member variables to hold state and initialize it however they want from 'initialize'.

The SimpleAggregateAdapter currently doesn't hold an instance of the user-defined simple UDAF class (i.e., it only creates instances of the AccumulatorType struct inside the UDAF class). We can change SimpleAggregateAdapter to hold an instance of the UDAF class if we want to allow authors to freely access member variables in it.

Reply:
> We can change SimpleAggregateAdapter to hold an instance of the UDAF class if we want to allow authors to freely access member variables in it.

This would be nice. Thanks.

Reply:
Hi @kagamiori, I will try to understand this method and see how it can be modified.

Reply:
Hi @liujiayi771, sorry for the delay. Here is a code pointer of how the simple scalar function interface calls the initialize() method: https://github.com/facebookincubator/velox/blob/ffc28ac1f16648356904bfea3af69ce8bfcd409a/velox/expression/SimpleFunctionAdapter.h#L213

The `fn_` here is a std::unique_ptr of the UDF class. Because this UDF instance is created, the author can add function-level states as member variables in the UDF class and the UDF's initialize() and call() member methods can access them directly. An example UDF that uses initialize() is below.
https://github.com/facebookincubator/velox/blob/ffc28ac1f16648356904bfea3af69ce8bfcd409a/velox/functions/lib/Re2Functions.h#L262

What @mbasmanova suggested is that we can do it similar in the SimpleAggregateAdapter so that the UDAF authors doesn't have to keep function-level states. Specifically, below is what I’m thinking:
1. The UDAF author doesn’t define a FunctionState struct, but rather add function-level variables as data members in the UDAF class (outside of its AccumulatorType struct).
2. The UDAF class has an initialize() method that receives the aggregation step, the types, and the constantInput, and assigns values to the data members in the UDAF class.
3. The AccumulatorType struct has a data member that is a pointer to the UDAF class. This would allow member methods inside AccumulatorType to access data members in the UDAF class. This UDAF-pointer can be set in SimpleAggregateAdapter::initializeNewGroupsInternal().

I’ll try to make a prototype to see if this works. Let’s discuss and review this design before coding in https://github.com/facebookincubator/velox/discussions/8711 first.

Reply:
Thanks for your reply @kagamiori. Let's discuss in #8711 further once the prototype has been validated for feasibility.

Reply:
Hi @kagamiori. Do you have time to make a prototype for this design. cc @rui-mo.